multiplexer/appd/run.go

```diff
@@ -203,7 +119,131 @@ func (a *Appd) Stop() error {
+			binaryPath = path
+			return filepath.SkipAll // Found it, stop searching
+		}
+		return nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to find executable binary in the archive: %w", err)
+	}
+	if binaryPath == "" {
+		return "", fmt.Errorf("no executable binary found in the archive for %s", version)
+	}
+	return binaryPath, nil
+}
+
+func decompressBinary(version string, binary []byte) error {
+	if isBinaryAlreadyDecompressed(version) {
+		return nil
+	}
+
+	// untar the binary.
+	gzipReader, err := gzip.NewReader(bytes.NewReader(binary))
+	if err != nil {
+		return fmt.Errorf("failed to read binary data for %s: %w", version, err)
+	}
+	defer gzipReader.Close()
+
+	targetDirectory := getDirectoryForVersion(version)
+	if err := os.MkdirAll(targetDirectory, 0o755); err != nil {
+		return fmt.Errorf("failed to create directory: %w", err)
+	}
+
+	// extract all files from the tar archive to the directory
+	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return fmt.Errorf("failed to read tar header: %w", err)
+		}
+
+		if header.FileInfo().IsDir() {
+			// Create directory
+			dirPath := filepath.Join(targetDirectory, header.Name)
+			if err := os.MkdirAll(dirPath, 0o755); err != nil {
+				return fmt.Errorf("failed to create directory %s: %w", dirPath, err)
+			}
+			continue
+		}
+
+		// Create file path
+		filePath := filepath.Join(targetDirectory, header.Name)
+
+		// Create parent directory if it doesn't exist
+		if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
+			return fmt.Errorf("failed to create parent directory for %s: %w", filePath, err)
+		}
+
+		// Create file
+		f, err := os.OpenFile(filePath, os.O_CREATE|os.O_RDWR, header.FileInfo().Mode())
+		if err != nil {
+			return fmt.Errorf("failed to create file %s: %w", filePath, err)
+		}
+
+		if _, err := io.Copy(f, tarReader); err != nil {
+			f.Close()
+			return fmt.Errorf("failed to copy file contents to %s: %w", filePath, err)
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// isBinaryAlreadyDecompressed returns true if the binary for the given version
+// has already been decompressed.
+func isBinaryAlreadyDecompressed(version string) bool {
+	dir := getDirectoryForVersion(version)
+	_, err := os.Stat(dir)
+	return err == nil
+}
+
+func verifyBinaryIsExecutable(pathToBinary string) error {
+	testCmd := exec.Command(pathToBinary, "--help")
+	testOutput, err := testCmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("binary validation failed (%s): %w\nOutput: %s", pathToBinary, err, string(testOutput))
+	}
+	return nil
+}
+
+// getDirectoryForCelestiaAppBinaries returns the directory where all
+// decompressed celestia-app binaries are stored. One directory exists per
+// version.
+func getDirectoryForCelestiaAppBinaries() string {
+	return filepath.Join(os.Getenv("HOME"), ".celestia-app", "bin")
```

Comment:
is this always correct? This should be the cli specified home flag right? I think there is an issue with that now also.

Reply:
I couldn't get the `--home` value plumbed here because the multiplexer gets created and added to the root command.

I will refactor this to use the CELESTIA_APP_HOME environment variable though.

Ref: https://github.com/celestiaorg/celestia-app?tab=readme-ov-file#environment-variables

Reply:
Got the `--home` value plumbed through.